### ISSUE TITLE
Unexpected SocketTimeoutException

### ISSUE THREAD

**Issue by mh-dm** [square/okhttp#567]
Using version: 1.3.0 release, obtained from http://mvnrepository.com/artifact/com.squareup.okhttp/okhttp

Stack trace:

```
Caused by: java.net.SocketTimeoutException
    at com.squareup.okhttp.internal.spdy.SpdyStream$SpdyDataInputStream.waitUntilReadable(SpdyStream.java:472)
    at com.squareup.okhttp.internal.spdy.SpdyStream$SpdyDataInputStream.read(SpdyStream.java:405)
    at com.squareup.okhttp.internal.http.UnknownLengthHttpInputStream.read(UnknownLengthHttpInputStream.java:39)
    at java.io.InputStream.read(InputStream.java:163)
    at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:198)
    at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:152)
    at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:167)
    at libcore.io.Streams.readFully(Streams.java:81)
    at java.io.DataInputStream.readInt(DataInputStream.java:124)
    etc ... But note that just readInt() was called.
```

The problem here is that a simple DataInputStream.readInt() triggers a SocketTimeoutException even though it shouldn't - the network is good (wifi) and the server is responsive.

Digging into this issue, I have localized the issue to code inside `com.squareup.okhttp.internal.spdy.SpdyStream.SpdyDataInputStream.receive(InputStream in, int byteCount)`:

```
  [...] // Some code here.

  // Fill the buffer without holding any locks. First fill [limit..buffer.length) if that
  // won't overwrite unread data. Then fill [limit..pos). We can't hold a lock, otherwise
  // writes will be blocked until reads complete.
  if (pos < limit) {
    int firstCopyCount = Math.min(byteCount, buffer.length - limit);
    Util.readFully(in, buffer, limit, firstCopyCount);
    limit += firstCopyCount;
    byteCount -= firstCopyCount;
    if (limit == buffer.length) {
      limit = 0;
    }
  }
  if (byteCount > 0) {
    Util.readFully(in, buffer, limit, byteCount);
    limit += byteCount;
  }

  synchronized (SpdyStream.this) {
    // Update the new limit, and mark the position as readable if necessary.
    this.limit = limit;
    if (this.pos == -1) {
      this.pos = firstNewByte;
      SpdyStream.this.notifyAll();
    }
  }
```

The problem is that `Util.readFully(in, buffer, limit, firstCopyCount);` can take a long time, say longer than 10s. This means that it takes that much until `SpdyStream.this.notifyAll();` is called. Furthermore, `com.squareup.okhttp.internal.spdy.SpdyStream.SpdyDataInputStream.waitUntilReadable()` is waiting on `notifyAll()` to be called. If it's not called after `readTimeoutMillis`(set to 10s) it times out and throws the unexpected SocketTimeoutException posted above. I've copied `waitUntilReadable()` below:

```
/**
 * Returns once the input stream is either readable or finished. Throws
 * a {@link SocketTimeoutException} if the read timeout elapses before
 * that happens.
 */
private void waitUntilReadable() throws IOException {
  long start = 0;
  long remaining = 0;
  if (readTimeoutMillis != 0) {
    start = (System.nanoTime() / 1000000);
    remaining = readTimeoutMillis;
  }
  try {
    while (pos == -1 && !finished && !closed && errorCode == null) {
      if (readTimeoutMillis == 0) {
        SpdyStream.this.wait();
      } else if (remaining > 0) {
        SpdyStream.this.wait(remaining);
        remaining = start + readTimeoutMillis - (System.nanoTime() / 1000000);
      } else {
        throw new SocketTimeoutException();
      }
    }
  } catch (InterruptedException e) {
    throw new InterruptedIOException();
  }
}
```

You might be thinking 'what's the problem?'. Nothing happened for 10s so a timeout makes sense. Except for the fact that stuff is happening, only in `Util.readFully(in, buffer, limit, firstCopyCount);`. That's where data is actually read (successfully, but slowly) from the input stream / network.

So if 30KB (realistic) are being read from the network, at 2KB/s, a SocketTimeoutException will always be thrown. Technically, this would be fine since the timeout is specified at 10s. However, note that only a simple readInt() was called but that resulted in 30KB being read from the network - this is an issue, which I'm not yet sure whether it is due to an issue inside okhttp or due to compression (say 30KB are required to decompress the first int).
Importantly, there is no way to distinguish between 'no network or server not responding' and 'connection is slow' since they are governed by the same timeout. Could that be added? (Or is there already a way to do this that I should know about?)

There is also another issue that revealed the issue above. On intel atom android device (thus single core, x86), on wifi, it sometimes (say 25% of time) takes 14s to read 30KB (all in Util.readFully) using okhttp (which leads to the SocketTimeoutException noted above). Before I investigate any deeper into okhttp - is there any busy wait/thread switching/thread starvation that could be happening inside okhttp that could potentially cause the slowness? (Yes, I'm that desperate for leads.)
